File: app/src/main/java/com/example/unitrade/ProductRepository.java
package com.example.unitrade;

import com.google.firebase.firestore.DocumentSnapshot;
import com.google.firebase.firestore.FirebaseFirestore;

import java.util.ArrayList;
import java.util.List;

public class ProductRepository {

    public interface ProductListCallback {
        void onSuccess(List<Product> products);

        void onFailure(Exception e);
    }

    public static void getActiveProductsByUser(String userId, ProductListCallback callback) {
        FirebaseFirestore.getInstance()
                .collection("products")

                .whereEqualTo("sellerId", userId)
                .get()
                .addOnSuccessListener(snapshot -> {
                    List<Product> products = new ArrayList<>();
                    for (DocumentSnapshot doc : snapshot) {
                        products.add(doc.toObject(Product.class));
                    }
                    callback.onSuccess(products);
                })
                .addOnFailureListener(callback::onFailure);
    }
}

File: app/src/main/java/com/example/unitrade/ProductRepository.kt
package com.example.unitrade

import io.github.jan.supabase.postgrest.postgrest
import kotlinx.coroutines.CoroutineScope
import kotlinx.coroutines.Dispatchers
import kotlinx.coroutines.launch
import kotlinx.coroutines.withContext

object ProductRepository {

    interface ProductListCallback {
        fun onSuccess(products: List<Product>)
        fun onFailure(e: Exception)
    }

    fun getActiveProductsByUser(userId: String, callback: ProductListCallback) {
        CoroutineScope(Dispatchers.IO).launch {
            try {
                val products = SupabaseClient.client.postgrest["products"]
                    .select {
                        eq("seller_id", userId)
                        eq("status", "active")
                    }
                    .decodeList<Product>()
                withContext(Dispatchers.Main) {
                    callback.onSuccess(products)
                }
            } catch (e: Exception) {
                withContext(Dispatchers.Main) {
                    callback.onFailure(e)
                }
            }
        }
    }
}